game-core/src/main/java/games/strategy/triplea/ai/pro/util/MoveBatcher.java

```diff
@@ -0,0 +1,164 @@
+
+  /**
+   * Starts a new sequence. This must be called before calling any add*() methods. A sequence
+   * indicates a logical dependency relationship between the moves in that sequence.
+   */
+  public void newSequence() {
+    moveSequences.add(new ArrayList<Move>());
+  }
+
+  /**
+   * Adds a transport load move. Ownership of the params is transferred to this object.
+   *
+   * @param unit The unit to move.
+   * @param route The route to move on.
+   * @param transport The transport to load.
+   */
+  public void addTransportLoad(final Unit unit, final Route route, final Unit transport) {
+    addMove(new Move(unit, route, transport));
+  }
+
+  /**
+   * Adds a move for the specified units. Ownership of the params is transferred to this object.
+   *
+   * @param units The units to move. Note that type is ArrayList explicitly to prevent an
+   *     unmodifiable List.of() being passed.
+   * @param route The route to move on.
+   */
+  public void addMove(final ArrayList<Unit> units, final Route route) {
+    addMove(new Move(units, route));
+  }
+
+  private void addMove(final Move newMove) {
+    final ArrayList<Move> sequence = moveSequences.get(moveSequences.size() - 1);
+    if (!sequence.isEmpty() && sequence.get(sequence.size() - 1).mergeWith(newMove)) {
+      return;
+    }
+    sequence.add(newMove);
+  }
+
+  /**
+   * Emits the accumulated moves to the passed output params.
+   *
+   * @param moveUnits The units to move.
+   * @param moveRoutes The routes to move along
+   * @param unitsToTransports The map from units to transports to be loaded.
+   */
+  public void batchAndEmit(
+      final List<Collection<Unit>> moveUnits,
+      final List<Route> moveRoutes,
+      final List<Map<Unit, Unit>> unitsToTransports) {
+    int i = 0;
+    for (final var sequence : moveSequences) {
+      if (!sequence.isEmpty()) {
+        mergeSequences(sequence, moveSequences.subList(i + 1, moveSequences.size()));
+        for (final Move move : sequence) {
+          move.addTo(moveUnits, moveRoutes, unitsToTransports);
+        }
+      }
+      i++;
+    }
+  }
+
+  /**
+   * Attempts to merge a sequence of moves with other sequences. Any sequence merged into this
+   * sequence will have its moves cleared.
+   *
+   * @param sequence The sequence to merge other sequences into.
+   * @param sequences The sequences to try to merge into the sequence.
+   */
+  private static void mergeSequences(
+      final ArrayList<Move> sequence, final List<ArrayList<Move>> sequences) {
+    for (final var otherSequence : sequences) {
+      boolean merge = (otherSequence.size() == sequence.size());
+      for (int j = 0; merge && j < sequence.size(); j++) {
+        merge = sequence.get(j).canMergeWith(otherSequence.get(j));
+      }
+      if (!merge) {
+        continue;
+      }
+      for (int j = 0; j < sequence.size(); j++) {
+        if (!sequence.get(j).mergeWith(otherSequence.get(j))) {
+          throw new RuntimeException("Unexpected");
```

Comment:
Ok, this is a bad message.
I'm someone that says that there's no such thing as an expected exception in general, because exception really are an extraordinary event per definition.
So 2 options here:
1. This is supposed to be an impossible condition, in this case throwing an `IlegalStateException` would be more appropriate here. It should contain a better message why this state is illegal though.
2. This exception might happen from time to time, in this case a simple message explaining what happened would be sufficient already.

Reply:
It's the first case, since earlier we call canMergeWith() on all the moves on line 151. Changed as you suggested.